src/address.cpp

```diff
@@ -0,0 +1,205 @@
+#include "veriblock/address.hpp"
+
+#include <cassert>
+
+#include "veriblock/base58.hpp"
+#include "veriblock/base59.hpp"
+#include "veriblock/hashutil.hpp"
+#include "veriblock/serde.hpp"
+
+namespace VeriBlock {
+
+constexpr const auto STARTING_CHAR = 'V';
+constexpr const auto MULTISIG_ENDING_CHAR = '0';
+
+constexpr const auto MULTISIG_ADDRESS_M_VALUE = 1;
+constexpr const auto MULTISIG_ADDRESS_N_VALUE = 2;
+constexpr const auto MULTISIG_ADDRESS_MIN_N_VALUE = 2;
+constexpr const auto MULTISIG_ADDRESS_MAX_N_VALUE = 58;
+constexpr const auto MULTISIG_ADDRESS_MAX_M_VALUE = 58;
+constexpr const auto MULTISIG_ADDRESS_DATA_END = 24;
+constexpr const auto MULTISIG_ADDRESS_CHECKSUM_END = 28;
+
+static std::string getDataPortionFromAddress(const std::string& address) {
+  assert(address.length() == ADDRESS_SIZE);
+  return address.substr(0, MULTISIG_ADDRESS_DATA_END + 1);
+}
+
+static bool isMultisig(const std::string& address) {
+  assert(address.length() == ADDRESS_SIZE);
+  return (address[ADDRESS_SIZE - 1] == MULTISIG_ENDING_CHAR);
+}
+
+static std::string getChecksumPortionFromAddress(const std::string& address,
+                                                 bool multisig) {
+  assert(address.length() == ADDRESS_SIZE);
+  if (multisig) {
+    return address.substr(MULTISIG_ADDRESS_DATA_END + 1,
+        MULTISIG_ADDRESS_CHECKSUM_END - MULTISIG_ADDRESS_DATA_END);
+  }
+  return address.substr(MULTISIG_ADDRESS_DATA_END + 1);
+}
+
+static bool isBase58String(const std::string& input) {
+  try {
+    auto decoded = DecodeBase58(input);
+    (void)decoded;
+    return true;
+  } catch (std::invalid_argument e) {
+    // do not throw - return status instead
+  }
+  return false;
+}
+
+static bool isBase59String(const std::string& input) {
+  try {
+    auto decoded = DecodeBase59(input);
+    (void)decoded;
+    return true;
+  } catch (std::invalid_argument e) {
+    // do not throw - return status instead
+  }
+  return false;
+}
+
+static std::string calculateChecksum(const std::string& data, bool multisig) {
+  std::vector<uint8_t> dataBytes = toBytes(data);
+  auto hash = sha256(dataBytes);
```

Comment:
no need to copy string to vector and then calculate sha256 from vector. you can calculate sha256 directly from string